Netbeans 352

SirIntellegence opened the pull request:
Moved changes from my master branch. Same as #393, but less of a mess on my end. I didn't see a way to fix this without opening a new pull request.

junichi11 commented:
Maybe I'm not a proper person to review this. So, I just would like to say about your git commits.

It would be nice if you can rebase & squash your commits yourself (see https://git-scm.com/book/id/v2/Git-Tools-Rewriting-History).

You should add the issue number([NetBeans-352]) to your commit message. (see other PRs and Contribution Guidelines (4): https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=74681408#SubmittingPullRequestonApacheNetBeans(incubating)-ContributionGuidelines)

Thanks for your work!

SirIntellegence commented:
Ya, these are commits I made beforehand. I moved the branch up through the changes with the "rebase" command. The changes already existed, I just changed what branch it was on. I plan on doing those things in the future though.

mdindoffer commented on debugger.jpda.projectsui/src/org/netbeans/modules/debugger/jpda/projectsui/BreakpointAnnotationProvider.java:
```diff
@@ -413,6 +413,9 @@ private void addAnnotationTo(JPDABreakpoint b, FileObject fo, int[] lines) {
         List<DebuggerBreakpointAnnotation> annotations = new ArrayList<DebuggerBreakpointAnnotation>();
         for (int l : lines) {
             try {
+                if (l - 1 < 0){
```
This does not seem to be related to the 352 issue, please remove.

mdindoffer commented on o.n.core/src/org/netbeans/core/ui/ProductInformationPanel.java:
```diff
@@ -44,7 +44,9 @@
 import javax.swing.event.HyperlinkListener;
 import org.netbeans.core.actions.HTMLViewAction;
+
 import static org.netbeans.core.ui.Bundle.*;
```
Move the static import up to a separate paragraph above the classic imports.

SirIntellegence replied:
Is there a setting for that in NetBeans?

mdindoffer commented on openide.modules/src/org/openide/modules/AboutMessageOverride.java:
```diff
@@ -0,0 +1,33 @@
+ * Register it with {@link ServiceProvider}
+ * @author astephens
+ */
+public interface AboutMessageOverride {
```
I think this may be considered as an API change? If so, it should be accompanied by a version bump and a quick API review process, right @JaroslavTulach?

JaroslavTulach replied:
Yes, this looks like an API change. In such case I suggest to follow http://wiki.netbeans.org/APIReviews (to certain extent). Thanks for CCing me, @mdindoffer!

JaroslavTulach commented on openide.modules/src/org/openide/modules/AboutMessageOverride.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public interface AboutMessageOverride {
+    public String formatAboutText(String bundleString, String productVersion,
+            String javaVersion, String vmVersion, String os, String encoding,
```
Are we introducing new interface to format a string? That would rather be handled with branding - see `NbBundle.getMessage`.

SirIntellegence replied:
See comment in the original pull request: https://github.com/apache/incubator-netbeans/pull/393#issuecomment-361332941
Note: when I mention `currentVersion`, I mean the System Property.

JaroslavTulach commented on o.n.core/src/org/netbeans/core/ui/ProductInformationPanel.java:
```diff
@@ -94,8 +98,30 @@
     public ProductInformationPanel() {
         initComponents();
+        descripTemplate = NbBundle.getMessage(ProductInformationPanel.class,
+                "LBL_description");
```
`getMessage` takes parameters - pass all the `(String bundleString, String productVersion, String javaVersion, String vmVersion, String os, String encoding, String locale, String userDir, String cacheDir, String updates, int fontSize, String javaRuntime)` in and let the one who does the branding format that somehow.

JaroslavTulach replied:
Enhancing branding this way is is also an API change, so you can still document it in `arch.xml` (I don't think there is `category="branding"`, if not, use `"property"`) and record it in `apichanges.xml` files + bump the version of the module.

neilcsmith-net commented:
@JaroslavTulach the actual issue here cannot be handled by branding as far as I can see - there needs to be a way to filter / change the arguments passed into `NbBundle.getMessage()` to do this.  As I mentioned on the first PR for this, I'm -1 to adding a new API to try and fix this one issue, because it's not the only place I've seen this problem.  I did suggest we provide a way to filter input arguments - https://github.com/apache/incubator-netbeans/pull/393#issuecomment-362662452  What would be your thoughts on trying to do that instead?

jtulach commented:
I am not sure what is the problem with _filtering_? As far as I can tell `MessageFormat` can ignore some arguments - isn't that the filtering?

The other part of the proposal is trying to dynamically change parameters of the format. That would break [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency). Rather than that, just submit a PR and add additional parameters to the `MessageFormat` call, in my opinion.

Should any form of such PR not be enough (and I don't know why), then there is always option to replace whole _About_ action and the dialog you any app based on top of NetBeans Platform.

neilcsmith-net commented:
Surely `NbBundle.getMessage()` isn't referentially transparent already?  At least with respect to locale and branding token, otherwise there wouldn't be a need for it.  Filter might have been a bad choice of word (think FilterInputStream) - but, yes, really I'm wanting a way to dynamically (well, at run time) transform the input arguments!

Taking a step back, branding cannot achieve what the OP wants to achieve because it's static, and here  there's a need to transform the input arguments into the message via code.  I've come across, and seen other people with, similar problems in various different areas of the platform, and sometimes hacked around them (including exactly what the OP is doing for this).  I was trying to come up with the least intrusive way we could fix it in one place, rather than PR's all over the platform and / or requiring people to rewrite big chunks of it.

They all come down to needing to change the input arguments to (or wrap entirely) `NbBundle.getMessage()`, _from code_.  Am I right in thinking there is no way to do that currently?  Do you have ideas on a better way to achieve this?

JaroslavTulach commented:
In a single execution the locale and branding token are assumed to be constant. Thus `NbBundle.getMessage()` runtime results are _effectively_ referentially transparent.

I believe the `java.text` formatting architecture supports pluggable formatters. Javadoc of [MessageFormat](https://docs.oracle.com/javase/7/docs/api/java/text/MessageFormat.html) mentions `choice` & co. Maybe you want to write another format and somehow plug it in?

neilcsmith-net commented:
To use pluggable formatters we'd still have to be able to wrap or override `NbBundle.getMessage()` to use our own `MessageFormat` to achieve this?

It's not a proposal to break the idea of referentially transparent you're talking about then - it's about being able to override / wrap the default behaviour of `NbBundle.getMessage()`.  That _could_ be abused, but the proposal is meant to be that you still get the same output for any distinct set of input arguments.

At the moment there is an assumption that a module is providing all / the right inputs you need to construct the output message you want, but this doesn't always hold in a platform application (it probably does for translations).  The OP's problem is one, but there are lots of times you might want to calculate something differently eg. turning an `X of Y files` message into an `X files remaining` message, etc.

JaroslavTulach commented:
Re. "there is an assumption that a module is providing ... the right inputs ... to construct ... message ... but this doesn't always hold in a platform application" - yes, that is the assumption. I'd like to cast permanent `-1` to change of this assumption. Please count my `-1` whenever such voting happens. I believe the right approach is to modify the underlaying module to provide more arguments to `NbBundle.getMessage` to make everybody happy. If nothing else, that at least increases the number of contributors to Apache NetBeans.

neilcsmith-net commented:
@JaroslavTulach fair enough.  I've obviously made a lousy case for why this is impossible to fix _inside_ the platform, sorry! :smile:

I've opened an issue in JIRA, linked to the original issue - not sure it's good discussing this on a PR.  Maybe we should close this and find consensus there?  Issues related to this do crop up now and again, but it would be good to have a solution that you can be in favour of!  And also one that solves @SirIntellegence problem.

https://issues.apache.org/jira/browse/NETBEANS-427

matthiasblaesing commented:
Is there still interest in this?

neilcsmith-net commented:
@matthiasblaesing well, I'd definitely still like to see us find a way to plug in new text and image branding plugins.  It's a real annoyance with the platform that customization is limited here.

matthiasblaesing commented:
My question was targetted at this PR, not so much on the general intention. This PR had its last action > 1,5 years(!) ago.

neilcsmith-net commented:
@matthiasblaesing true!  I don't think this is likely to be merged in the direction it's currently going, so shall we bring up above discussion elsewhere and close?  The problem it's trying to solve is valid, but I would like to tidy up some of these old PRs as we move towards next release.

SirIntellegence commented:
It would likely be a nightmare to merge at the moment... I am not proficient enough with git to know how that would work after all of the things were moved around... I don't really currently have any ideas other than what is here, so feel free to close it and move the discussion elsewhere.

JaroslavTulach commented:
I am going through open pull requests and checking if they can be merged. This one is distracting my attention. Closing. Open when you have a request for review.

matthiasblaesing commented:
@JaroslavTulach you could just filter on "is:open -label:work-in-progress" - that way the label has a real meaning
